c++/src/BloomFilter.cc

```diff
@@ -0,0 +1,337 @@
+
+    uint64_t nb = optimalNumOfBits(expectedEntries, fpp);
+    // make 'mNumBits' multiple of 64
+    mNumBits = nb + (BITS_OF_LONG - (nb % BITS_OF_LONG));
+    mNumHashFunctions = optimalNumOfHashFunctions(expectedEntries, mNumBits);
+    mBitSet.reset(new BitSet(mNumBits));
+  }
+
+  void BloomFilterImpl::addBytes(const char * data, int64_t length) {
+    uint64_t hash64 = getBytesHash(data, length);
+    addHash(hash64);
+  }
+
+  void BloomFilterImpl::addLong(int64_t data) {
+    addHash(getLongHash(static_cast<uint64_t>(data)));
+  }
+
+  bool BloomFilterImpl::testBytes(const char * data, int64_t length) const {
+    uint64_t hash64 = getBytesHash(data, length);
+    return testHash(hash64);
+  }
+
+  bool BloomFilterImpl::testLong(int64_t data) const {
+    return testHash(getLongHash(static_cast<uint64_t>(data)));
+  }
+
+  uint64_t BloomFilterImpl::sizeInBytes() const {
+    return getBitSize() >> SHIFT_3_BITS;
+  }
+
+  uint64_t BloomFilterImpl::getBitSize() const {
+    return mBitSet->bitSize();
+  }
+
+  uint64_t BloomFilterImpl::getNumHashFunctions() const {
+    return mNumHashFunctions;
+  }
+
+  DIAGNOSTIC_PUSH
+
+#if defined(__clang__)
+  DIAGNOSTIC_IGNORE("-Wundefined-reinterpret-cast")
+#endif
+
+#if defined(__GNUC__)
+  DIAGNOSTIC_IGNORE("-Wstrict-aliasing")
+#endif
+
+  // caller should make sure input proto::BloomFilter is valid since
+  // no check will be performed in the following constructor
+  BloomFilterImpl::BloomFilterImpl(const proto::BloomFilter& bloomFilter) {
+    mNumHashFunctions = bloomFilter.numhashfunctions();
+
+    const std::string& bitsetStr = bloomFilter.utf8bitset();
+    mNumBits = bitsetStr.size() << SHIFT_3_BITS;
+    checkArgument(mNumBits % BITS_OF_LONG == 0, "numBits should be multiple of 64!");
+
+    const uint64_t * bitset = reinterpret_cast<const uint64_t *>(bitsetStr.data());
+    if (isLittleEndian()) {
+      mBitSet.reset(new BitSet(bitset, mNumBits));
+    } else {
+      std::vector<uint64_t> longs(mNumBits >> SHIFT_6_BITS);
+      for (size_t i = 0; i != longs.size(); ++i) {
+        // convert little-endian to big-endian
+        const uint64_t src = bitset[i];
+        uint64_t& dst = longs[i];
+        for (size_t bit = 0; bit != 64; bit += 8) {
+          dst |= (((src & (0xFFu << bit)) >> bit) << (56 - bit));
+        }
+      }
+
+      mBitSet.reset(new BitSet(longs.data(), mNumBits));
+    }
+  }
+
+  void BloomFilterImpl::addDouble(double data) {
+    addLong(reinterpret_cast<int64_t&>(data));
+  }
+
+  bool BloomFilterImpl::testDouble(double data) const{
+    return testLong(reinterpret_cast<int64_t&>(data));
+  }
+
+  DIAGNOSTIC_POP
+
+  void BloomFilterImpl::addHash(uint64_t hash64) {
+    int32_t hash1 = static_cast<int32_t>(hash64 & 0xffffffff);
+    int32_t hash2 = static_cast<int32_t>(hash64 >> 32);
+
+    for (int32_t i = 1; i <= static_cast<int32_t>(mNumHashFunctions); ++i) {
+      int32_t combinedHash = hash1 + i * hash2;
```

Comment:
should this be uint64_t to get more hash space? This is being casted to `uint64_t` anyway below.

Reply:
I think combinedHash is deliberately designed to use 32 bits according to the "Bloom Filter Index" section in https://orc.apache.org/specification/ORCv1/. Not sure if it will have any problem if I change it to uint64_t